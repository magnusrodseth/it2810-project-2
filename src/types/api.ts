export type ApiCommitsOptions = {
  id: string
  withStats?: boolean
  perPage?: number
}

export type ApiIssuesOptions = {
  id: string
}

export type AuthenticationResponse = {
  ok: boolean
  message: string
}

export type CommitStats = {
  additions: number
  deletions: number
  total: number
}

export type Commit = {
  id: string
  short_id: string
  title: string
  author_name: string
  author_email: string
  authored_date: string
  committer_name: string
  committer_email: string
  committed_date: string
  created_at: string
  message: string
  parent_ids: string[]
  web_url: string
  stats?: CommitStats
}

export type Issue = {
  id: number
  issue_type: 'issue' | 'incident' | 'test_case'
  labels: string[]
  state: 'all' | 'opened' | 'closed'
}
