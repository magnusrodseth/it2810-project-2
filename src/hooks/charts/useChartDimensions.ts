import { Dimensions } from '../../types/dimensions'
import { breakpoints } from '../../utils/common/breakpoints'
import useWindowDimensions from '../common/useWindowDimensions'

/**
 * Calculates the dimensions of the graph based on the window dimensions.
 *
 * @returns {Dimensions} The dimensions of the graph.
 **/
const useChartDimensions = (): Dimensions => {
  const { width, height } = useWindowDimensions()

  if (width <= breakpoints.small) {
    return { width: width * 0.75, height: width * 0.75 }
  }

  if (width <= breakpoints.medium) {
    return { width: width * 0.8, height: height * 0.6 }
  }

  if (width <= breakpoints.large) {
    return { width: width * 0.7, height: height * 0.6 }
  }

  return { width: width * 0.6, height: height * 0.6 }
}

export default useChartDimensions
