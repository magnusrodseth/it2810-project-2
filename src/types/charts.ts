import { CommitStats } from './api'

export type LineChartEntry = {
  timestamp: number
  commits: number
  author?: string
  stats?: CommitStats
}

export type BarChartEntry = {
  label: string
  count: number
}

export type LineChartData = LineChartEntry[]
export type BarChartData = BarChartEntry[]

export type YAxisLabel = keyof Pick<LineChartEntry, 'commits'>

export type Position = {
  left: string
  right: string
  top: string
  bottom: string
}

export type ChartType = 'commits' | 'issues'

export type Contributor = string

export type Option = {
  label: string
  value: string
}
