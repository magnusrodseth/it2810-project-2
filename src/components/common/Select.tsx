import React, { ChangeEvent } from 'react'
import { Option, Contributor } from '../../types/charts'

interface SelectProps {
  label: string
  options: Option[]
  selected?: Contributor
  onChange: (e: ChangeEvent<HTMLSelectElement>) => void
}

const Select: React.FC<SelectProps> = ({ label, options, onChange, selected }) => {
  return (
    <div className="form-control w-full max-w-xs">
      <select
        className="select select-bordered"
        value={selected == undefined ? 'default' : selected}
        onChange={onChange}>
        <option disabled value="default">
          {label}
        </option>
        {options.map((option) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
    </div>
  )
}

export default Select
