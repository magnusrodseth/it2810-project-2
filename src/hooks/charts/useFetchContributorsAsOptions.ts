import { useEffect, useState } from 'react'
import { Option } from '../../types/charts'
import useGetContributorsAsOptionsQuery from '../queries/useGetContributorsQuery'

/**
 * Fetches contributors as options on initial render.
 **/
const useFetchContributorsAsOptions = () => {
  const [options, setOptions] = useState<Option[]>([])

  useEffect(() => {
    (async () => {
      setOptions(await useGetContributorsAsOptionsQuery())
    })()
  }, [])

  return options
}

export default useFetchContributorsAsOptions
