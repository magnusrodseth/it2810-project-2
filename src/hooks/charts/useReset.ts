import { Dispatch, SetStateAction, useContext } from 'react'
import { AppContext } from '../../state/context'
import { Position } from '../../types/charts'

/**
 * Resets the graph to its default position.
 **/
const useReset = (
  [position, setPosition]: [Position, Dispatch<SetStateAction<Position>>],
  [positionReferences, setPositionReferences]: [Position, Dispatch<SetStateAction<Position>>]
) => {
  const { setContributor, setWithStats } = useContext(AppContext)

  const reset = () => {
    setPosition({
      ...position,
      left: 'dataMin',
      right: 'dataMax',
      top: 'dataMax+1',
      bottom: 'dataMin'
    })

    setPositionReferences({
      ...positionReferences,
      left: '',
      right: '',
      top: 'dataMax+50'
    })

    // Resets to default state
    setContributor(undefined)
    setWithStats(false)
  }

  return reset
}

export default useReset
