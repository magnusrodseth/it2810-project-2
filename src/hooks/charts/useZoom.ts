import { Dispatch, SetStateAction } from 'react'
import { LineChartData, Position } from '../../types/charts'
import getYAxisDomain from '../../utils/charts/getYAxisDomain'

/**
 * Zooms the graph to the selected area.
 *
 * Inspiration: https://codesandbox.io/s/l4pq6x00xq
 **/
const useZoom = (
  [position, setPosition]: [Position, Dispatch<SetStateAction<Position>>],
  [positionReferences, setPositionReferences]: [Position, Dispatch<SetStateAction<Position>>],
  data?: LineChartData
) => {
  const zoom = () => {
    let { left: refLeft, right: refRight } = positionReferences

    if (refLeft === refRight || refRight === '') {
      setPositionReferences({
        ...positionReferences,
        left: refLeft,
        right: refRight
      })

      return
    }

    if (refLeft > refRight) {
      // prettier-ignore
      [refLeft, refRight] = [refRight, refLeft]
    }

    if (!data) {
      return
    }

    const [bottom, top] = getYAxisDomain(data, 'commits', 1)
    const [refBottom, refTop] = getYAxisDomain(data, 'commits', 1)

    setPosition({
      ...position,
      left: refLeft,
      right: refRight,
      bottom: bottom.toString(),
      top: top.toString()
    })

    setPositionReferences({
      ...positionReferences,
      left: '',
      right: '',
      bottom: refBottom.toString(),
      top: refTop.toString()
    })
  }

  return zoom
}

export default useZoom
