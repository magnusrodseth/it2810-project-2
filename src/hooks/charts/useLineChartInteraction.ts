import { useState } from 'react'
import { LineChartData, Position } from '../../types/charts'
import useReset from './useReset'
import useZoom from './useZoom'

const initialPosition: Position = {
  left: 'dataMin',
  right: 'dataMax',
  top: 'dataMax+1',
  bottom: '0'
}

const initialPositionReferences: Position = {
  left: '',
  right: '',
  top: 'dataMax+20',
  bottom: '0'
}

/**
 * Wrapper hook for all interaction with the line chart.
 *
 * This hook handles the positions, the position references when zooming,
 * the zooming itself, and resetting the chart.
 **/
const useLineChartInteraction = (data?: LineChartData) => {
  const [position, setPosition] = useState(initialPosition)
  const [positionReferences, setPositionReferences] = useState(initialPositionReferences)

  const zoom = useZoom([position, setPosition], [positionReferences, setPositionReferences], data)

  const reset = useReset(
    [position, setPosition],
    [positionReferences, setPositionReferences]
  )

  return {
    position,
    setPosition,
    positionReferences,
    setPositionReferences,
    zoom,
    reset
  }
}

export default useLineChartInteraction
