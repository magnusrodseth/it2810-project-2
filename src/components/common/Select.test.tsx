import Select from './Select'
import renderer from 'react-test-renderer'
import { Option } from '../../types/charts'

describe('Select', () => {
  it('renders correctly', () => {
    const options: Option[] = [
      {
        label: 'label',
        value: 'value'
      }
    ]

    const component = renderer.create(
      <Select label="Test" options={options} onChange={() => { }} selected={undefined} />
    )

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
