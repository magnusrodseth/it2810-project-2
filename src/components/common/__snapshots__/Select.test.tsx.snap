// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders correctly 1`] = `
<div
  className="form-control w-full max-w-xs"
>
  <select
    className="select select-bordered"
    onChange={[Function]}
    value="default"
  >
    <option
      disabled={true}
      value="default"
    >
      Test
    </option>
    <option
      value="value"
    >
      label
    </option>
  </select>
</div>
`;
